File: spice2/MNASolution.cpp
//
// Created by rhys on 18/04/2021.
//

#include <iostream>
#include "MNASolution.h"

MNASolution::MNASolution(std::map<int, double> nodeVoltages, std::vector<MNAElement *> elements) {
    this->nodeVoltages = nodeVoltages;
    this->elements = elements;
}

bool MNASolution::approxEquals(MNASolution otherMnaSolution) {
    std::vector<int> keys;
    std::vector<int> otherKeys;

    for(auto n : nodeVoltages){
        keys.push_back(n.first);
    }

    for(auto n : otherMnaSolution.nodeVoltages){
        otherKeys.push_back(n.first);
    }

    for(auto key : keys){
        if(!numApproxEquals(getNodeVoltage(key), otherMnaSolution.getNodeVoltage(key))){
            return false;
        }
    }

    if((!hasAllCurrents(otherMnaSolution))){
        return false;
    }

    if(!otherMnaSolution.hasAllCurrents(*this)) {
        return false;
    }

    return true;
}

bool MNASolution::numApproxEquals(double a, double b) {
    return std::abs(a-b) < 1e-6;
}

double MNASolution::getNodeVoltage(int nodeIndex) {
    return nodeVoltages.at(nodeIndex);
}

double MNASolution::getCurrentForResistor(MNAElement resistor) {
    return -getVoltage(resistor) / resistor.value;
}

double MNASolution::getVoltage(MNAElement element) {
    return nodeVoltages.at(element.n1) - nodeVoltages.at(element.n0);
}

bool MNASolution::hasAllCurrents(MNASolution mnaSolution) {
    for(auto e : mnaSolution.elements){
        if(!hasMatchingElement(e)){
            return false;
        }
    }

    return true;
}

bool MNASolution::hasMatchingElement(MNAElement *element) {
    for(auto e : elements){
        if(e->n0 == element->n0 && e->n1 == element->n1 && numApproxEquals(e->currentSolution, element->currentSolution)){
            return true;
        }
    }

    return false;
}

File: spice3/MNASolution.py
from math import fabs

from MNAElement import MNAElement

def num_approx_equals(a, b) -> bool:
    return fabs(a-b) < 1E-6

class MNASolution:
    def __init__(self, node_voltages: dict, elements: list):
        self.node_voltages: dict = node_voltages
        self.elements = elements

    def approx_equals(self, solution) -> bool:
        keys = self.node_voltages.keys()
        other_keys = solution.node_voltages.keys()

        key_difference = [x for x in keys if x not in other_keys]
        assert(len(key_difference) == 0)

        for k in keys:
            if not num_approx_equals(self.get_node_voltage(k), solution.get_node_voltage(k)):
                return False

        if not self.has_all_currents(solution):
            return False

        if not solution.has_all_currents(self):
            return False

        return True

    def has_all_currents(self, solution):
        for e in solution.elements:
            if not self.has_matching_element(e):
                return False
        return True

    def has_matching_element(self, element):
        for e in self.elements:
            if e.n0 == element.n0 and e.n1 == element.n1 and num_approx_equals(e.current_solution, element.current_solution):
                return True

        return False

    def get_current_for_resistor(self, resistor: MNAElement):
        v = - self.get_voltage(resistor)
        r = resistor.value
        i = v / r

        return i

    def get_node_voltage(self, node_index):
        return self.node_voltages[node_index]

    def get_voltage(self, element: MNAElement):
        return self.node_voltages[element.n1] - self.node_voltages[element.n0]
